cpp/oneapi/dal/algo/decision_forest/backend/gpu/train_splitter_impl_dpc.cpp

```diff
@@ -621,485 +127,517 @@ train_splitter_impl<Float, Bin, Index, Task, use_private_mem>::compute_random_sp
 
-    sycl::event last_event;
-
-    last_event = queue.submit([&](sycl::handler& cgh) {
+    sycl::event last_event = queue.submit([&](sycl::handler& cgh) {
         cgh.depends_on(deps);
-        local_accessor_rw_t<hist_type_t> local_hist_buf(local_hist_buf_size, cgh);
+        local_accessor_rw_t<byte_t> local_byte_buf(local_buf_byte_size, cgh);
+
+        cgh.parallel_for(nd_range, [=](sycl::nd_item<2> item) {
+            auto sbg = item.get_sub_group();
+            const Index node_idx = item.get_global_id(1);
+            if (node_idx > (node_count - 1)) {
+                return;
+            }
 
-        cgh.parallel_for(nd_range, [=](sycl::nd_item<1> item) {
-            const Index node_idx = item.get_global_id()[0];
-            const Index node_id = node_indices_ptr[node_ind_ofs + node_idx];
+            const Index node_id = node_idx;
             Index* node_ptr = node_list_ptr + node_id * impl_const_t::node_prop_count_;
 
+            const Index local_id = item.get_local_id(0);
+
+            const Index row_ofs = node_ptr[impl_const_t::ind_ofs];
+            const Index row_count = node_ptr[impl_const_t::ind_lrc];
+
             split_smp_t sp_hlp;
+            // Check node impurity
+            if (!sp_hlp.is_valid_impurity(node_imp_list_ptr, node_id, imp_threshold, row_count)) {
+                return;
+            }
+            split_info_t bs;
 
-            Index bs_ftr_bin = impl_const_t::leaf_mark_;
-            Index bs_ftr_id = impl_const_t::leaf_mark_;
-            Index bs_left_count = 0;
+            // slm pointers declaration
+            byte_t* local_byte_buf_ptr = local_byte_buf.get_pointer().get();
+            hist_type_t* local_hist_buf_ptr =
+                get_buf_ptr<hist_type_t>(&local_byte_buf_ptr, local_hist_buf_size);
+            Float* local_buf_float_ptr = get_buf_ptr<Float>(&local_byte_buf_ptr, local_size);
 
-            hist_type_t* local_hist_buf_ptr = local_hist_buf.get_pointer().get();
+            bs.init_clear(item, local_hist_buf_ptr + 0 * hist_prop_count, hist_prop_count);
 
-            hist_type_t* bs_left_hist =
-                fill_zero(local_hist_buf_ptr + 0 * hist_prop_count, hist_prop_count);
+            for (Index ftr_idx = 0; ftr_idx < selected_ftr_count; ftr_idx++) {
+                split_info_t ts;
+                ts.init(local_hist_buf_ptr + 1 * hist_prop_count, hist_prop_count);
+                ts.ftr_id = selected_ftr_list_ptr[node_id * selected_ftr_count + ftr_idx];
+                const Index id =
+                    (local_id < row_count) ? tree_order_ptr[row_ofs + local_id] : index_max;
+                const Index bin =
+                    (local_id < row_count) ? data_ptr[id * column_count + ts.ftr_id] : index_max;
+                const Float response = (local_id < row_count) ? response_ptr[id] : Float(0);
+                const Index response_int =
+                    (local_id < row_count) ? static_cast<Index>(response) : -1;
```

Comment:
In context of this code it is really doesn't matter, all values out of range [0, class_count - 1] will be ignored. So in case of unsigned type it will be MAX_INT-1 or MAX_LONG-1 depending on type template.